service/s3/s3crypto/crypto_mode.go

```diff
@@ -0,0 +1,138 @@
+package s3crypto
+
+import (
+	"encoding/base64"
+	"fmt"
+	"io"
+
+	"github.com/aws/aws-sdk-go/aws/awserr"
+)
+
+// CryptoMode is an abstraction layer that deals with encryption of
+// the contents and which key provider to use.
+type CryptoMode interface {
+	EncryptContents(io.Writer, io.Reader) error
+	DecryptContents(KeyProvider, io.ReadCloser) (io.ReadCloser, error)
```

Comment:
Looking over this i think it makes sense for `EncryptContents` to follow the same pattern as `DecryptContents`. This puts it more in line with the stream patterns used by the standard library such as `compress/gzip`. The `io.Copy` pattern is kind of unique where as the stream pattern is more general.

```go
EncryptContents(io.Reader) (io.Reader, error)
```